R/misc.R

```diff
@@ -436,13 +436,36 @@ check_type <- function(dat, quant = TRUE, types = NULL, call = caller_env()) {
   }
 
   if (!all(all_good)) {
-    rlang::abort(
-      paste0(
-        "All columns selected for the step",
-        " should be ",
-        label,
-        "."
-      ),
+    info <- get_types(dat)[!all_good, ]
+    classes <- map_chr(info$type, function(x) x[1])
+    counts <- vctrs::vec_split(info$variable, classes)
+    counts$count <- lengths(counts$val)
+    counts$text_len <- cli::console_width() - 19 - (counts$count > 1) -
+      nchar(counts$key)
+
+    # cli::ansi_collapse() doesn't work for length(x) == 1
+    # https://github.com/r-lib/cli/issues/590
+    variable_collapse <- function(x, width) {
+      x <- paste0("{.var ", x, "}")
+      if (length(x) == 1) {
+        res <- cli::ansi_strtrim(x, width = width)
+      } else {
+        res <- cli::ansi_collapse(x, width = width, style = "head")
+      }
+      res
+    }
+
+    problems <- glue::glue_data(
+      counts,
+      "{count} {key} variable{ifelse(count == 1, '', 's')} \\
+      found: {purrr::map2_chr(val, text_len, variable_collapse)}"
```

Comment:
I believe we can write out `variable_collapse()` altogether, instead writing:

```suggestion
      found: {paste0('{.var ', val, '}')}"
```

This would resolve the current issue where a comma is placed in two-item lists.

This does indeed mean that the single, long variable name is not truncated, instead line-wrapped to the console. I'm okay with that edge case for the sake of reduced complexity.

Reply:
The suggested change results in:

![Screenshot 2023-10-11 at 9 47 39 AM](https://github.com/tidymodels/recipes/assets/14034784/c7c23773-3efc-4f47-b804-68ec93f4393e)

I wrote `variable_collapse()` because I wanted to use `ansi_collapse()` so we only print out one line of variables instead of all of them

Reply:
I removed the comma for two item list now! in [`9106ffd` (#1235)](https://github.com/tidymodels/recipes/pull/1235/commits/9106ffd0be561ee04acb253669346afcc0886a3f)

``` r
library(recipes)
library(modeldata)

recipe(~., data = iris) |>
  step_date(Sepal.Length, Sepal.Width, Species) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date or datetime.
#> • 2 double variables found: `Sepal.Length` and `Sepal.Width`
#> • 1 factor variable found: `Species`

recipe(~., data = ames) |>
  step_date(all_predictors()) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date or datetime.
#> • 40 factor variables found: `MS_SubClass`, `MS_Zoning`, `Street`, …
#> • 12 double variables found: `Lot_Frontage`, `Mas_Vnr_Area`, …
#> • 22 integer variables found: `Lot_Area`, `Year_Built`, …
```

Reply:
Heard! I'm with it, then.